README.md
# [rede_pos](https://pub.dev/packages/rede_pos_pg)

A plugin wrapper to use rede pos hardware sdk in your flutter apps.

**ONLY WORKS ON ANDROID**

## Usage

### Configure your project

Change **minSdkVersion to 19** in your app
```
 defaultConfig {
        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
        applicationId "com.example.test_rede_pos_package"
        minSdkVersion 19
        targetSdkVersion 28
        versionCode flutterVersionCode.toInteger()
        versionName flutterVersionName
        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
    }
``` 

### Usage

Import the RedePosPg class. Use the availble static methods.

```
// Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String transactionResults = "";
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      // payment: R$90,00 in 3x
      var operationStatus = await RedePosPg.startPayment(PaymentType.CREDITO_PARCELADO, 9000, installments: 3);      
      transactionResults += "payment: R\$90,00 in 3x: [$operationStatus]\n";
      // payment: R$1,00
      operationStatus = await RedePosPg.startPayment(PaymentType.CREDITO_A_VISTA, 100);
      transactionResults += "payment: R\$1,00: [$operationStatus]\n";
      // reversal
      operationStatus = await RedePosPg.startReversal();
      transactionResults += "reversal: [$operationStatus]\n";
      operationStatus = await RedePosPg.startReprint() ? "Success" :  "Failure";
      transactionResults += "reprint: [$operationStatus]\n";
    } on PlatformException {
      transactionResults = 'Failed.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _results = transactionResults;
    });
  }
```

CHANGELOG.md
## 0.0.1

* Setup

## 0.0.2

* Payments
* Reversal
* Reprint

## 0.0.3

* bug fix

## 0.0.4

* cast bug fix

test/rede_pos_pg_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:rede_pos_pg/rede_pos_pg.dart';

void main() {
  const MethodChannel channel = MethodChannel('rede_pos_pg');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await RedePosPg.startReversal(), '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:rede_pos_pg_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/rede_pos_pg.dart
import 'dart:async';

import 'package:flutter/services.dart';

enum PaymentType {
  CREDITO_A_VISTA,
  CREDITO_PARCELADO,
  CREDITO_PARCELADO_EMISSOR,
  DEBITO,
  VOUCHER,
}

class RedePosPg {
  static const MethodChannel _channel = const MethodChannel('rede_pos_pg');

  /// Starts a new payment. The ammount is the transaction value multiplied by 100.
  /// Returns the payment status code:
  /// AUTHORIZED: Transação realizada/estornada com sucesso
  /// FAILED: Falha na transação
  /// DECLINED: Transação negada.
  static Future<String> startPayment(PaymentType paymentType, int ammount, {int installments = 0}) async {
    final String result = await _channel.invokeMethod('payment', {
      'paymentType': paymentType.toString().replaceAll("PaymentType.", ""),
      'ammount': ammount,
      'installments': installments
    });
    return result;
  }

  /// Cancel a payment.
  /// Returns the reversal status code:
  /// AUTHORIZED: Transação realizada/estornada com sucesso
  /// FAILED: Falha na transação
  /// DECLINED: Transação negada.
   static Future<String> startReversal() async {
    final String result = await _channel.invokeMethod('reversal');
    return result;
  }

  /// Reprint a payment.
  /// Returns true if printed. Otherwise false.  
   static Future<bool> startReprint() async {
    final bool result = await _channel.invokeMethod('reprint');
    return result;
  }
}


android/src/main/java/io/zafeapps/rede_pos_pg/RedePosPgPlugin.java
package io.zafeapps.rede_pos_pg;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;

import java.util.Objects;
import java.util.logging.Level;
import java.util.logging.Logger;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import rede.smartrede.sdk.FlexTipoPagamento;
import rede.smartrede.sdk.Payment;
import rede.smartrede.sdk.PaymentIntentBuilder;
import rede.smartrede.sdk.RedePaymentValidationError;
import rede.smartrede.sdk.RedePayments;

/**
 * RedePosPgPlugin
 */
public class RedePosPgPlugin implements MethodCallHandler, PluginRegistry.ActivityResultListener {

    private static final int PAYMENT_REQUEST_CODE = 1001;
    private static final int REVERSAL_REQUEST_CODE = 1002;
    private static final int REPRINT_REQUEST_CODE = 1003;

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "rede_pos_pg");
        RedePosPgPlugin plugin = new RedePosPgPlugin(registrar.activity());
        registrar.addActivityResultListener(plugin);
        channel.setMethodCallHandler(plugin);
    }

    private Result result;
    private Activity activity;
    private RedePayments redePayments;

    private RedePosPgPlugin(Activity activity) {
        this.activity = activity;
        this.redePayments = RedePayments.getInstance(activity);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        this.result = result;
        mountPayment(call);
        doReversal(call);
        reprint(call);
    }

    private void reprint(MethodCall call) {
        if (call.method.equals("reprint")) {
            try {
                Intent reprint = redePayments.intentForReprint();
                this.activity.startActivityForResult(reprint, REPRINT_REQUEST_CODE);
            } catch (ActivityNotFoundException e) {
                Logger.getLogger(this.getClass().getName()).log(Level.SEVERE, e.getMessage());
                this.result.error("Error", e.getMessage(), null);
            }
        }
    }

    private void mountPayment(MethodCall call) {
        if (call.method.equals("payment")) {
            try {
                FlexTipoPagamento paymentType = FlexTipoPagamento.valueOf(call.<String>argument("paymentType"));
                Long ammount = Long.parseLong(Objects.requireNonNull(call.argument("ammount")).toString());
                int installments = Integer.parseInt(Objects.requireNonNull(call.argument("installments")).toString());
                PaymentIntentBuilder paymentIntentBuilder = redePayments.intentForPaymentBuilder(paymentType, ammount);
                if (installments > 1) paymentIntentBuilder.setInstallments(installments);
                Intent intent = paymentIntentBuilder.build();
                this.activity.startActivityForResult(intent, PAYMENT_REQUEST_CODE);
            } catch (ActivityNotFoundException | RedePaymentValidationError e) {
                Logger.getLogger(this.getClass().getName()).log(Level.SEVERE, e.getMessage());
                this.result.error("Error", e.getMessage(), null);
            }
        }
    }

    private void doReversal(MethodCall call) {
        if (call.method.equals("reversal")) {
            try {
                Intent reversal = redePayments.intentForReversal();
                this.activity.startActivityForResult(reversal, REVERSAL_REQUEST_CODE);
            } catch (ActivityNotFoundException e) {
                Logger.getLogger(this.getClass().getName()).log(Level.SEVERE, e.getMessage());
                this.result.error("Error", e.getMessage(), null);
            }
        }
    }


    @Override
    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        if (handlerPaymentResponse(requestCode, resultCode, data)) return true;
        else if (handlerReversalResponse(requestCode, resultCode, data)) return true;
        return handlerReprintResponse(requestCode, resultCode, data);
    }

    private boolean handlerReprintResponse(int requestCode, int resultCode, Intent data) {
        if (requestCode == REPRINT_REQUEST_CODE) {
            try {
                if (resultCode == Activity.RESULT_OK) {
                    this.result.success(true);
                } else {
                    this.result.error("Error on reprint", null, null);
                }
            } catch (Exception e) {
                this.result.error("Error:", e.getMessage(), null);
            }
        }
        return false;
    }

    private boolean handlerReversalResponse(int requestCode, int resultCode, Intent data) {
        if (requestCode == REVERSAL_REQUEST_CODE) {
            try {
                if (resultCode == Activity.RESULT_OK) {
                    if (data != null) {
                        Payment payment = RedePayments.getPaymentFromIntent(data);
                        this.result.success(Objects.requireNonNull(payment, "payment not found").getStatus().toString());
                        return true;
                    }
                } else {
                    this.result.error("Error on reversal operation", null, null);
                }
            } catch (Exception e) {
                this.result.error("Error:", e.getMessage(), null);
            }
        }
        return false;
    }

    private boolean handlerPaymentResponse(int requestCode, int resultCode, Intent data) {
        if (requestCode == PAYMENT_REQUEST_CODE) {
            try {
                if (resultCode == Activity.RESULT_OK) {
                    if (data != null) {
                        Payment payment = RedePayments.getPaymentFromIntent(data);
                        this.result.success(Objects.requireNonNull(payment, "payment not found").getStatus().toString());
                        return true;
                    }
                } else if (resultCode == Activity.RESULT_CANCELED) {
                    this.result.error("Payment cancelled by the user", null, null);
                }
            } catch (Exception e) {
                this.result.error("Error:", e.getMessage(), null);
            }
        }
        return false;
    }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:rede_pos_pg/rede_pos_pg.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _results = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String transactionResults = "";
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      // payment: R$90,00 in 3x
      var operationStatus = await RedePosPg.startPayment(PaymentType.CREDITO_PARCELADO, 9000, installments: 3);      
      transactionResults += "payment: R\$90,00 in 3x: [$operationStatus]\n";
      // payment: R$1,00
      operationStatus = await RedePosPg.startPayment(PaymentType.CREDITO_A_VISTA, 100);
      transactionResults += "payment: R\$1,00: [$operationStatus]\n";
      // reversal
      operationStatus = await RedePosPg.startReversal();
      transactionResults += "reversal: [$operationStatus]\n";
      operationStatus = await RedePosPg.startReprint() ? "Success" :  "Failure";
      transactionResults += "reprint: [$operationStatus]\n";
    } on PlatformException {
      transactionResults = 'Failed.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _results = transactionResults;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on: $_results\n'),
        ),
      ),
    );
  }
}
